src/amberscript/parser.cc

```diff
@@ -370,19 +372,42 @@ Result Parser::ParseShaderBlock() {
 
   shader->SetFormat(format);
 
-  r = ValidateEndOfStatement("SHADER command");
-  if (!r.IsSuccess())
-    return r;
+  token = tokenizer_->PeekNextToken();
+  if (token->IsIdentifier() && token->AsString() == "VIRTUAL_FILE") {
+    tokenizer_->NextToken();  // Skip VIRTUAL_FILE
+
+    token = tokenizer_->NextToken();
+    if (!token->IsIdentifier() && !token->IsString())
+      return Result("expected virtual file path after VIRTUAL_FILE");
 
-  std::string data = tokenizer_->ExtractToNext("END");
-  if (data.empty())
-    return Result("SHADER must not be empty");
+    r = ValidateEndOfStatement("SHADER command");
+    if (!r.IsSuccess())
+      return r;
 
-  shader->SetData(data);
+    auto path = token->AsString();
 
-  token = tokenizer_->NextToken();
-  if (!token->IsIdentifier() || token->AsString() != "END")
-    return Result("SHADER missing END command");
+    std::string data;
+    r = script_->GetVirtualFiles()->Get(path, &data);
```

Comment:
What about just `script_->GetVirtualFile(path, &data);` keep the implementation hidden

Reply:
`VirtualFileStore` is handed down to the compilers in places the full script is not available. Also gives me a clean place to add the virtual file tests, without cluttering the script further.

Reply:
That's fine, and `Get` and `Add` make sense as the API into the virtual file system. I'm just thinking if the parser should care, or if it should just tell the script has files to add/get.

Reply:
Done.